### DISCUSSION TITLE
How can I set default value of a field in pocketbase

### DISCUSSION THREAD

**Discussion by syedwalihasan** [pocketbase/pocketbase#2193] [Q&A] [answered]
I want to set guid for the record as a default value automatically generated like we have 'id' on User collection.
How can I do it in pocketbase collection. I tried looking into documentation but was not successful 

<img width="200" alt="image" src="https://user-images.githubusercontent.com/18749369/229280891-3d24672d-cef9-4236-9eed-6d7766c2cb6e.png">

<img width="509" alt="image" src="https://user-images.githubusercontent.com/18749369/229280735-86e53eec-3c15-459b-a6ac-afc15df59ebf.png">

**Comment by ganigeorgiev** [comment, marked as answer]
Defaults can be set only client-side with the record submission OR if you want to do it server-side, you'll have to [use PocketBase as framework](https://pocketbase.io/docs/use-as-framework/) and attach to the [`OnModelBefore*` hooks](https://pocketbase.io/docs/event-hooks/#db-hooks).

**Comment by syedwalihasan** [comment]
And I believe to set a field to auto increment also, we need to use pocketbase as  framework ?
I checked below discussion
https://github.com/pocketbase/pocketbase/discussions/1038

**Reply by ganigeorgiev** [reply]
Yes, there is no autoincrement field support out of the box.

**Reply by syedwalihasan** [reply]
thanks for providing the direction @ganigeorgiev ! :)

**Comment by nickheyer** [comment]
This seems like it would be a very common sense feature

**Reply by Buzut** [reply]
I concur! Especially for Select fields.

**Comment by capodisseny** [comment]
Hi, I'm trying to set on server side with pb_hooks, 
the hook is running but when i do this the creation fails:
```

// fires for every db model
onModelBeforeCreate((e) => {

     e.model.label = "Some label"

})
```

How could be implemented? and also how can I get the request auth id here? so i can do
```
// fires for every db model
onModelBeforeCreate((e) => {

     e.model.owner = e.request.auth.id

})

```

thanks!

**Reply by ganigeorgiev** [reply]
The dynamic record fields can be accessed or changed only through their designated get/set methods - https://pocketbase.io/docs/js-records/#getset-record-fields.

> and also how can I get the request auth id here?

You can't. You have access to the request context only in the `*Request` suffixed hooks (this is made more clear with the upcoming v0.23.0 APIs and new documentation).

**Reply by ganigeorgiev** [reply]
Since I forgot to mention that the authenticated user could be also an admin (superuser in v0.23.0), here is a short example how to extract the regular auth record id:

```js
onRecordBeforeCreateRequest((e) => {
    // https://pocketbase.io/docs/js-routing/#retrieving-the-current-auth-state
    const info = $apis.requestInfo(e.httpContext);

    if (info.admin) {
        return null // skip for admins
    }

    if (!info.authRecord) {
        throw new UnauthorizedError()
    }

    e.record.set("owner", info.authRecord.id)
}, "yourCollection")
```

With v0.23.0 the above could be written like this:

```js
// with v0.23.0 this also runs BEFORE the validations,
// so you can mark the "owner" relation field as non-empty from the collection field options
// and it will do the necessary checks for you
onRecordCreateRequest((e) => {
    if (e.hasSuperuserAuth()) {
        return e.next() // skip for superusers
    }

    e.record.set("owner", e.auth?.id)

    e.next();
}, "yourCollection")
```

**Reply by capodisseny** [reply]
Perfect ! thanks @ganigeorgiev  🙌
